spark2/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead.java

```diff
@@ -0,0 +1,388 @@
+    File checkpoint = new File(parent, "checkpoint");
+    Table table = createTable(location.toString());
+
+    DataSourceOptions options = new DataSourceOptions(ImmutableMap.of(
+        "path", location.toString(),
+        "checkpointLocation", checkpoint.toString()));
+    IcebergSource source = new IcebergSource();
+
+    StreamingReader streamingReader = (StreamingReader) source.createMicroBatchReader(
+        Optional.empty(), checkpoint.toString(), options);
+
+    List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+    Collections.reverse(snapshotIds);
+    long initialSnapshotId = snapshotIds.get(0);
+
+    // Getting all appends from initial snapshot.
+    List<MicroBatch> pendingBatches = streamingReader.getChangesWithRateLimit(
+        initialSnapshotId, 0, true, false, Long.MAX_VALUE);
+    Assert.assertEquals(pendingBatches.size(), 4);
+
+    List<Long> batchSnapshotIds = pendingBatches.stream()
+        .map(MicroBatch::snapshotId)
+        .collect(Collectors.toList());
+    Assert.assertEquals(batchSnapshotIds, snapshotIds);
+
+    // Getting appends from initial snapshot with index, 1st snapshot will be filtered out.
+    List<MicroBatch> pendingBatches1 = streamingReader.getChangesWithRateLimit(
+        initialSnapshotId, 1, true, false, Long.MAX_VALUE);
+
+    Assert.assertEquals(pendingBatches1.size(), 4);
+    MicroBatch batch = pendingBatches1.get(0);
+    Assert.assertEquals(batch.sizeInBytes(), 0L);
+    Assert.assertEquals(batch.endFileIndex(), 1);
+    Assert.assertTrue(Iterables.isEmpty(batch.tasks()));
+
+    // Getting appends from 2nd snapshot, 1st snapshot should be filtered out.
+    long snapshotId2 = snapshotIds.get(1);
+    List<MicroBatch> pendingBatches2 = streamingReader.getChangesWithRateLimit(
+        snapshotId2, 0, false, false, Long.MAX_VALUE);
+
+    Assert.assertEquals(pendingBatches2.size(), 3);
+    List<Long> batchSnapshotIds1 = pendingBatches2.stream()
+        .map(MicroBatch::snapshotId)
+        .collect(Collectors.toList());
+    Assert.assertEquals(batchSnapshotIds1.indexOf(initialSnapshotId), -1);
+
+    // Getting appends from last snapshot with index, should have no task included.
+    long lastSnapshotId = snapshotIds.get(3);
+    List<MicroBatch> pendingBatches3 = streamingReader.getChangesWithRateLimit(
+        lastSnapshotId, 1, false, false, Long.MAX_VALUE);
+
+    Assert.assertEquals(pendingBatches3.size(), 1);
+    MicroBatch batch1 = pendingBatches3.get(0);
+    Assert.assertEquals(batch1.sizeInBytes(), 0L);
+    Assert.assertEquals(batch1.endFileIndex(), 1);
+    Assert.assertTrue(Iterables.isEmpty(batch.tasks()));
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testGetChangesWithRateLimit() throws IOException {
+    File parent = temp.newFolder("test");
+    File location = new File(parent, "table");
+    File checkpoint = new File(parent, "checkpoint");
+    Table table = createTable(location.toString());
```

Comment:
I think the MicroBatch tests are easier to understand because they use fake files that are all the same size. Could you convert the tests for `getChangesWithRateLimit` to do the same thing?